src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc

```diff
@@ -0,0 +1,682 @@
+}
+
+//////////////////////////////////////////////////
+void OpticalTactilePluginPrivate::DepthCameraCallback(
+        const ignition::msgs::PointCloudPacked &_msg)
+{
+  // This check avoids running the callback at t=0 and getting
+  // unexpected markers in the scene
+  if (!this->initialized)
+      return;
+
+  this->cameraMsg = _msg;
+  this->newCameraMsg = true;
+}
+
+//////////////////////////////////////////////////
+void OpticalTactilePluginPrivate::UnpackPointCloudMsg(
+                        const ignition::msgs::PointCloudPacked &_msg)
+{
+  // 3D vector to store XYZ points
+  this->imageXYZ = std::vector<std::vector<std::vector<float>>>
+      (_msg.width(),
+      std::vector<std::vector<float>>(_msg.height(),
+      std::vector<float>(3)));
+
+  // Unpack point cloud message
+  std::string msgBuffer = _msg.data();
+  char *msgBufferIndex = msgBuffer.data();
+
+  for (uint32_t j = 0; j < _msg.height(); ++j)
+  {
+    for (uint32_t i = 0; i < _msg.width(); ++i)
+    {
+      int fieldIndex = 0;
+
+      // x coordinate in pixel (i,j)
+      this->imageXYZ[i][j][0] =  *reinterpret_cast<float *>(
+        msgBufferIndex + _msg.field(fieldIndex++).offset());
+      // y coordinate in pixel (i,j)
+      this->imageXYZ[i][j][1] = *reinterpret_cast<float *>(
+        msgBufferIndex + _msg.field(fieldIndex++).offset());
+      // z coordinate in pixel (i,j)
+      this->imageXYZ[i][j][2] = *reinterpret_cast<float *>(
+        msgBufferIndex + _msg.field(fieldIndex++).offset());
+
+      msgBufferIndex += _msg.point_step();
+    }
+  }
+}
+
+//////////////////////////////////////////////////
+void OpticalTactilePluginPrivate::ComputeNormalForces(
+                    const ignition::msgs::PointCloudPacked &_msg,
+                    const bool _visualizeForces)
+{
+  // 2D vector to store normal force vectors
+  std::vector<std::vector<ignition::math::Vector3f>> normalForces(_msg.width(),
+      std::vector<ignition::math::Vector3f>(_msg.height()));
+
+  // Visualization is downsampled according to plugin parameter <resolution>
+  // if <visualize_forces> is set to true
+  bool visualizeRow = true;
+  int rowCounter = 0;
+
+  bool visualizeColumn = true;
+  int columnCounter = 0;
+
+  int index = 1;
+
+
+  for (uint32_t j = 1; j < (_msg.height() - 1); ++j)
+  {
+    for (uint32_t i = 1; i < (_msg.width() - 1); ++i)
+    {
+      // Compute normal forces
+      double dxdi =
+          (this->imageXYZ[i+1][j][0] -
+          this->imageXYZ[i-1][j][0])/
+          std::abs(this->imageXYZ[i+1][j][1] -
+          this->imageXYZ[i-1][j][1]);
+
+      double dxdj =
+          (this->imageXYZ[i][j+1][0]
+          - this->imageXYZ[i][j-1][0])/
+          std::abs(this->imageXYZ[i][j+1][2]
+          - this->imageXYZ[i][j-1][2]);
+
+      ignition::math::Vector3f direction(-1, -dxdi, -dxdj);
+
+      // todo(anyone) multiply vector by contact forces info
+      normalForces[i][j] = direction.Normalized();
+
+      if (!_visualizeForces)
```

Comment:
I guess right now there isn't a reason to compute the normals at all if visualizeForce == false. You can add a todo that the normals will be published. That can be in a later PR as this one is already pretty long.